File: 032/cpp/main.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int m = 0;
        vector<int> stack;
        int k = -1;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                stack.push_back(i);
            } else {
                if (stack.empty()) {
                    k = i;
                    continue;
                }
                stack.pop_back();
                int j = !stack.empty() ? stack.back() : k;
                m = std::max(m, i - j);
            }
        }
        return m;
    }
};

File: 032/rust/main.rs
use std::cmp;

impl Solution {
    pub fn longest_valid_parentheses(s: String) -> i32 {
        let mut m = 0;
        let mut stack = Vec::new();
        let mut k = -1;
        let s = s.as_bytes();
        for i in 0..s.len() {
            let c = s[i] as char;
            if c == '(' {
                stack.push(i as i32);
            } else {
                if stack.is_empty() {
                    k = i as i32;
                    continue;
                }
                stack.pop();
                let j = if !stack.is_empty() { *stack.last().unwrap() } else { k };
                m = cmp::max(m, (i as i32) - j); 
            }
        }
        return m;
    }
}
